### PR TITLE
Add Scene Background Texture

### PR DESCRIPTION
**Pull request by gz65555**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)

Add scene background texture mode, you can simply add a texture as scene background:

``` typescript
scene.background.mode = BackgroundMode.Texture;
scene.background.textureFillMode = BackgroundTextureFillMode.FitWidth;
scene.background.texture = texture;
```

### CODE REVIEW

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 1]
```diff
@@ -205,6 +211,27 @@ export class BasicRenderPipeline {
     }
   }
 
+  private _drawBackgroundTexture(engine: Engine, background: Background) {
+    const rhi = engine._hardwareRenderer;
+    const { shaderData, shader, renderState } = engine._backgroundTextureMaterial;
+    const { _backgroundTextureMesh, canvas } = engine;
+
+    if (
+      (this._lastCanvasWidth !== canvas.width || this._lastCanvasHeight !== canvas.width) &&
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:220]: You can use `Vector2` instead two `number` to save pointer memory.

File: packages/core/src/enums/BackgroundTextureFillMode.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+   */
+  FitWidth,
+  /**
+   * Adapt to height, scale based on texture height.
+   */
+  FitHeight,
+  /**
+   * Scale mode, Scale the texture completely according to the canvas size.
+   */
+  Scale
```
[the remark this answers is not in the document]
Comment (gz65555) [reply packages/core/src/enums/BackgroundTextureFillMode.ts:13]: `Zoom` maybe a better option.

File: packages/core/src/Background.ts [hunk 3]
```diff
@@ -25,4 +28,90 @@ export class Background {
...
+   * Background texture fill mode.
+   * @remarks When `mode` is `BackgroundMode.Texture`, the property will take effects.
+   * @defaultValue `BackgroundTextureFillMode.FitHeight`
+   */
+  public get textureFillMode(): BackgroundTextureMode {
+    return this._textureFillMode;
+  }
+
+  public set textureFillMode(value: BackgroundTextureMode) {
+    if (value === this._textureFillMode) {
```
Comment (GuoLei1990) [inline packages/core/src/Background.ts:63]:
```typescript
if ( this._textureFillMode!==value)
```

File: packages/core/src/enums/BackgroundTextureMode.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+  FitWidth,
+  /**
+   * Adapt to height, scale based on texture height.
+   */
+  FitHeight,
+  /**
+   * Scale mode, Scale the texture completely according to the canvas size.
+   */
+  ScaleToFill
+}
```
Comment (GuoLei1990) [inline packages/core/src/enums/BackgroundTextureMode.ts:14]:
```typescript
/**
 * Filling mode of background texture.
 */
export enum BackgroundTextureMode {
  /* Maintain the aspect ratio and scale the texture to fit the width of the canvas. */
  AspectFitWidth,
  /* Maintain the aspect ratio and scale the texture to fit the height of the canvas. */
  AspectFitHeight,
  /* Scale the texture so it exactly fills the canvas, the aspect ratio may be adjusted if necessary. */
  Fill
}
```

File: packages/core/src/Background.ts [hunk 5]
```diff
@@ -25,4 +28,90 @@ export class Background {
...
+      this._textureFillMode = value;
+      this._resizeBackgroundTexture();
+    }
+  }
+
+  /**
+   * Constructor of Background.
+   * @param _engine Engine Which the background belongs to.
+   */
+  constructor(private _engine: Engine) {}
```
Comment (GuoLei1990) [inline packages/core/src/Background.ts:73]: add @internal for  constructor comments.

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 6]
```diff
@@ -205,6 +211,29 @@ export class BasicRenderPipeline {
...
+  private _drawBackgroundTexture(engine: Engine, background: Background) {
+    const rhi = engine._hardwareRenderer;
+    const { shaderData, shader, renderState } = engine._backgroundTextureMaterial;
+    const { _backgroundTextureMesh, canvas } = engine;
+
+    if (
+      (this._lastCanvasWidth !== canvas.width || this._lastCanvasHeight !== canvas.height) &&
+      background._textureFillMode !== BackgroundTextureMode.ScaleToFill
+    ) {
+      this._lastCanvasWidth = canvas.width;
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:223]: Use Vector2 to save last size to save pointer memory.

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 7]
```diff
@@ -205,6 +211,29 @@ export class BasicRenderPipeline {
     }
   }
 
+  private _drawBackgroundTexture(engine: Engine, background: Background) {
+    const rhi = engine._hardwareRenderer;
+    const { shaderData, shader, renderState } = engine._backgroundTextureMaterial;
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:216]: `shaderData`, `shader`, `renderState` just use once, cahe is not good!

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 8]
```diff
@@ -205,6 +210,28 @@ export class BasicRenderPipeline {
     }
   }
 
+  private _drawBackgroundTexture(engine: Engine, background: Background) {
+    const rhi = engine._hardwareRenderer;
+    const { _backgroundTextureMaterial } = engine;
+    const { _backgroundTextureMesh, canvas } = engine;
```
Comment (GuoLei1990) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:216]:
```typescript
const { _backgroundTextureMaterial } = engine;
const { _backgroundTextureMesh, canvas } = engine;
```
can merge one
```
const { _backgroundTextureMaterial,  _backgroundTextureMesh, canvas } = engine;
```

File: packages/core/src/Background.ts [hunk 9]
```diff
@@ -25,4 +28,91 @@ export class Background {
...
+  public get textureFillMode(): BackgroundTextureMode {
+    return this._textureFillMode;
+  }
+
+  public set textureFillMode(value: BackgroundTextureMode) {
+    if (value !== this._textureFillMode) {
+      this._textureFillMode = value;
+      this._resizeBackgroundTexture();
+    }
+  }
```
Comment (singlecoder) [inline packages/core/src/Background.ts:68]: default is public

File: packages/core/src/Background.ts [hunk 10]
```diff
@@ -25,4 +28,91 @@ export class Background {
...
+  /**
+   * @internal
+   */
+  _resizeBackgroundTexture(): void {
+    if (!this._texture) {
+      return;
+    }
+    const { canvas } = this._engine;
+    const { width, height } = canvas;
+    const { width: textureWidth } = this._texture;
```
Comment (singlecoder) [inline packages/core/src/Background.ts:85]: textureWidth just use once, not need cache

and in ScaleToFill not used

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 11]
```diff
@@ -205,6 +210,28 @@ export class BasicRenderPipeline {
...
+
+    if (
+      (this._lastCanvasSize.x !== canvas.width || this._lastCanvasSize.y !== canvas.height) &&
+      background._textureFillMode !== BackgroundTextureMode.ScaleToFill
+    ) {
+      this._lastCanvasSize.setValue(canvas.width, canvas.height);
+      background._resizeBackgroundTexture();
+    }
```
Comment (singlecoder) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:224]: canvas.width and canvas.height should cache

File: packages/core/src/Background.ts (hunk 10 above)
Comment (GuoLei1990) [reply packages/core/src/Background.ts:85]: +1

File: packages/core/src/Background.ts (hunk 9 above)
Comment (GuoLei1990) [reply packages/core/src/Background.ts:68]: +1

**Review by GuoLei1990 — APPROVED** [review]
+1

**Review by cptbtptpbcptdtptp — APPROVED** [review]
+1

**Pull request merged**